hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/listener/ClientSmartListenerService.java

```diff
@@ -20,103 +20,112 @@
-        @Override
-        public Set<ClientEventRegistration> createNew(ClientRegistrationKey arg) {
-            return Collections.newSetFromMap(new ConcurrentHashMap<ClientEventRegistration, Boolean>());
-        }
-    };
-    private final ConcurrentMap<ClientRegistrationKey, Set<ClientEventRegistration>> registrations
-            = new ConcurrentHashMap<ClientRegistrationKey, Set<ClientEventRegistration>>();
-    private final Object regMutex = new Object();
+    private final Set<Member> members = new HashSet<Member>();
+    private final Map<ClientRegistrationKey, Map<Address, ClientEventRegistration>> registrations
+            = new ConcurrentHashMap<ClientRegistrationKey, Map<Address, ClientEventRegistration>>();
+    private final Object listenerRegLock = new Object();
 
-    public ClientSmartListenerService(HazelcastClientInstanceImpl client, int eventThreadCount, int eventQueueCapacity) {
+    public ClientSmartListenerService(HazelcastClientInstanceImpl client,
+                                      int eventThreadCount, int eventQueueCapacity) {
         super(client, eventThreadCount, eventQueueCapacity);
+        client.getClientClusterService().addMembershipListener(this);
     }
 
     @Override
     public String registerListener(ListenerMessageCodec codec, EventHandler handler) {
-        ClientMessage request = codec.encodeAddRequest(true);
-
-        Collection<Member> members = client.getClientClusterService().getMemberList();
-
         String userRegistrationId = UuidUtil.newUnsecureUuidString();
-        ClientRegistrationKey registrationKey = new ClientRegistrationKey(userRegistrationId, request, handler, codec);
-        for (Member member : members) {
-            invoke(registrationKey, member.getAddress());
+        synchronized (listenerRegLock) {
+
+            ClientRegistrationKey registrationKey = new ClientRegistrationKey(userRegistrationId, handler, codec);
+            registrations.put(registrationKey, new ConcurrentHashMap<Address, ClientEventRegistration>());
+            try {
+                for (Member member : this.members) {
+                    invoke(registrationKey, member.getAddress());
+                }
+            } catch (Exception e) {
+                deregisterListener(userRegistrationId);
+                throw new HazelcastException("Listener can not be added", e);
+            }
+            return userRegistrationId;
         }
-        return userRegistrationId;
     }
 
-    private void invoke(ClientRegistrationKey registrationKey, Address address) {
-        ClientMessage request = registrationKey.getRequest();
+    private void invoke(ClientRegistrationKey registrationKey, Address address) throws Exception {
+        ListenerMessageCodec codec = registrationKey.getCodec();
+        ClientMessage request = codec.encodeAddRequest(true);
         EventHandler handler = registrationKey.getHandler();
         handler.beforeListenerRegister();
```

Comment:
This callbacks are mainly used for cache invalidation. This is the safer approach to not have invalid near cache. We could argue for better solution later. 

>> This will also mean that the events will get fired again when a new member is added
I did not understand what you mean by this. Do you mean ,for example,  entryUpdated can be fired twice for same event when a new node joined ?